opm/flowdiagnostics/DerivedQuantities.cpp

```diff
@@ -0,0 +1,239 @@
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#if HAVE_CONFIG_H
+#include <config.h>
+#endif // HAVE_CONFIG_H
+
+#include <opm/flowdiagnostics/DerivedQuantities.hpp>
+#include <numeric>
+
+namespace Opm
+{
+namespace FlowDiagnostics
+{
+
+    /// The F-Phi curve.
+    ///
+    /// The F-Phi curve is an analogue to the fractional flow
+    /// curve in a 1D displacement. It can be used to compute
+    /// other interesting diagnostic quantities such as the Lorenz
+    /// coefficient. For a technical description see Shavali et
+    /// al. (SPE 146446), Shook and Mitchell (SPE 124625).
+    Graph flowCapacityStorageCapacityCurve(const Toolbox::Forward& injector_solution,
+                                           const Toolbox::Reverse& producer_solution,
+                                           const std::vector<double>& pv)
+    {
+        const auto& ftof = injector_solution.fd.timeOfFlight();
+        const auto& rtof = producer_solution.fd.timeOfFlight();
+        if (pv.size() != ftof.size() || pv.size() != rtof.size()) {
+            throw std::runtime_error("flowCapacityStorageCapacityCurve(): "
+                                     "Input solutions must have same size.");
+        }
+
+        // Sort according to total travel time.
+        const int n = pv.size();
+        typedef std::pair<double, double> D2;
+        std::vector<D2> time_and_pv(n);
+        for (int ii = 0; ii < n; ++ii) {
+            time_and_pv[ii].first = ftof[ii] + rtof[ii]; // Total travel time.
+            time_and_pv[ii].second = pv[ii];
+        }
+        std::sort(time_and_pv.begin(), time_and_pv.end());
+
+        // Compute Phi.
+        std::vector<double> Phi(n + 1);
+        Phi[0] = 0.0;
+        for (int ii = 0; ii < n; ++ii) {
+            Phi[ii+1] = time_and_pv[ii].second;
+        }
+        std::partial_sum(Phi.begin(), Phi.end(), Phi.begin());
+        const double vt = Phi.back(); // Total pore volume.
+        for (int ii = 1; ii < n+1; ++ii) { // Note limits of loop.
+            Phi[ii] /= vt; // Normalize Phi.
+        }
+
+        // Compute F.
+        std::vector<double> F(n + 1);
+        F[0] = 0.0;
+        for (int ii = 0; ii < n; ++ii) {
+            F[ii+1] = time_and_pv[ii].second / time_and_pv[ii].first;
+        }
+        std::partial_sum(F.begin(), F.end(), F.begin());
+        const double ft = F.back(); // Total flux.
+        for (int ii = 1; ii < n+1; ++ii) { // Note limits of loop.
+            F[ii] /= ft; // Normalize Phi.
+        }
```

Comment:
The code to compute `F` and `Phi` is essentially equal.  Any reason not to factor that out to a helper--e.g.,
```C++
template <class InputValues, class ExtractElement>
std::vector<double>
cumulativeQuantitiy(const InputValues& input,
                    ExtractElement&& extract)
{
    auto q = std::vector<double>{};
    q.reserve(input.size() + 1);
    q.push_back(0.0);

    for (const auto& e : input) {
        q.push_back(extract(e));
    }

    std::partial_sum(q.begin(), q.end(), q.begin());

    const auto t = q.back();
    for (auto& qi : q) { qi /= t; }

    return q;
}
```
whence computing `Phi` and `F` becomes
```C++
const auto Phi = cumulativeQuantity(time_and_pv, [](const D2& i) { return i.first; });
const auto F = cumulativeQuantity(time_and_pv, [](const D2& i) { return i.second / i.first; });
```